%YAML 1.2
# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
#
# NVIDIA CORPORATION and its licensors retain all intellectual property
# and proprietary rights in and to this software, related documentation
# and any modifications thereto.  Any use, reproduction, disclosure or
# distribution of this software and related documentation without an express
# license agreement from NVIDIA CORPORATION is strictly prohibited.
---
name: sync
components:
- name: left_image_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: left_image_receiver
    min_size: 1
- name: right_image_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: right_image_receiver
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
    policy: 0
- name: left_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: left_cam_receiver
    min_size: 1
- name: right_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: right_cam_receiver
    min_size: 1
- type: nvidia::isaac_ros::Synchronization
  parameters:
    left_image_input: left_image_receiver
    right_image_input: right_image_receiver
    left_cam_input: left_cam_receiver
    right_cam_input: right_cam_receiver
    left_image_output: left_image_transmitter
    right_image_output: right_image_transmitter
    left_cam_output: left_cam_transmitter
    right_cam_output: right_cam_transmitter
    inputs: [left_image_receiver, right_image_receiver, left_cam_receiver, right_cam_receiver]
    outputs: [left_image_transmitter, right_image_transmitter, left_cam_transmitter, right_cam_transmitter]
- name: left_image_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: left_image_transmitter
    min_size: 1
- name: right_image_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: right_image_transmitter
    min_size: 1
- name: left_cam_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: left_cam_transmitter
    min_size: 1
- name: right_cam_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: right_cam_transmitter
    min_size: 1
---
name: ess
components:
- name: data_receiver_left
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_left
    min_size: 1
- name: data_receiver_right
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver_right
    min_size: 1
- name: adapter
  type: nvidia::cvcore::tensor_ops::ImageAdapter
  parameters:
    message_type: VideoBuffer
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 12
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter
    min_size: 1
- name: pool
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 1
    block_size: 18432000
    num_blocks: 40
- type: nvidia::cvcore::ESS
  parameters:
    output_name: disparity
    pool: pool
    left_image_receiver: data_receiver_left
    right_image_receiver: data_receiver_right
    output_transmitter: data_transmitter
    output_adapter: adapter
    image_type: RGB_U8
    pixel_mean: [-128, -128, -128]
    normalization: [0.00392156862, 0.00392156862, 0.00392156862]
    standard_deviation: [0.5, 0.5, 0.5]
    max_batch_size: 1
    input_layer_width: 960
    input_layer_height: 576
    model_input_type: RGB_U8
    engine_file_path: engine_file_path_placeholder
    input_layers_name: [input_left, input_right]
    output_layers_name: [output_left]
    preprocess_type: RESIZE
---
name: disparity_compositor
components:
- name: disparity_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: disparity_in
    min_size: 1
- name: left_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: left_cam_receiver
    min_size: 1
- name: right_cam_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: right_cam_receiver
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter
    min_size: 1
- name: pool
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 1
    block_size: 18432000
    num_blocks: 40
- type: nvidia::isaac_ros::DisparityCompositor
  parameters:
    left_camera_model_receiver: left_cam_receiver
    right_camera_model_receiver: right_cam_receiver
    disparity_receiver: disparity_in
    output_transmitter: data_transmitter
    min_disparity: 0
    max_disparity: -1
---
name: vault
components:
- name: signal
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
    policy: 0
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: signal
    min_size: 1
- name: vault
  type: nvidia::gxf::Vault
  parameters:
    source: signal
    max_waiting_count: 1
    drop_waiting: false
---
name: connections
components:
- type: nvidia::gxf::Connection
  parameters:
    source: sync/left_image_transmitter
    target: ess/data_receiver_left
- type: nvidia::gxf::Connection
  parameters:
    source: sync/right_image_transmitter
    target: ess/data_receiver_right
- type: nvidia::gxf::Connection
  parameters:
    source: sync/left_cam_transmitter
    target: disparity_compositor/left_cam_receiver
- type: nvidia::gxf::Connection
  parameters:
    source: sync/right_cam_transmitter
    target: disparity_compositor/right_cam_receiver
- type: nvidia::gxf::Connection
  parameters:
    source: ess/data_transmitter
    target: disparity_compositor/disparity_in
- type: nvidia::gxf::Connection
  parameters:
    source: disparity_compositor/data_transmitter
    target: vault/signal
---
components:
- name: clock
  type: nvidia::gxf::RealtimeClock
- type: nvidia::gxf::GreedyScheduler
  parameters:
    clock: clock
    stop_on_deadlock: false
